osbrain/tests/test_agent.py

```diff
@@ -105,6 +105,88 @@ def test_agent_shutdown(nsproxy):
     assert 'a0' not in nsproxy.list()
 
 
+def test_close_socket_agent_shutdown(nsproxy):
+    """
+    Check that the socket is closed and the socket file removed after the agent
+    is shut down
+    """
+    agent = run_agent('name')
+    address = agent.bind('PUSH')
+    agent.shutdown()
+    assert not address.address.exists()
+
+
+def test_close_socket_agent_kill(nsproxy):
+    """
+    Check that the socket is closed and the socket file removed after the agent
+    is killed
+    """
+    agent = run_agent('name')
+    address = agent.bind('PUSH')
+    agent.kill()
+    assert not address.address.exists()
+
+
+def test_close_socket_agent_blocked(nsproxy):
+    """
+    Check that the socket is closed and the socket file removed when a blocked
+    agent is requested to shut down
+    """
+    def log(agent, msg):
+        time.sleep(3)
+        return msg
+
+    alice = run_agent('alice')
+    bob = run_agent('bob')
+
+    addr = alice.bind('REP', alias='main', handler=log)
+    bob.connect(addr, alias='main')
+
+    Timer(1, nsproxy.shutdown).start()
+    bob.send('main', 'Hello world')
+    rep = bob.recv('main')
+    print(rep)
+
+    assert not addr.address.exists()
```

Comment:
I don't think this test is correct. If you are able to `print(rep)` it means the agent is not blocked. The idea here would be to make sure the agent is blocked when we call for `nsproxy.shutdown()` (and that the shutdown succeeds and the socket file gets closed).

Something like:

```python
def test_close_socket_agent_blocked(nsproxy):                                   
    """                                                                         
    Check that the socket is closed and the socket file removed when a blocked  
    agent is requested to shut down.                                            
    """                                                                         
    def block(agent):                                                           
        agent.send('out', 'blocking...')                                        
                                                                                
    blocker = run_agent('blocker')                                              
    blocker.set_method(block)                                                   
    addr = blocker.bind('PUSH', alias='out')                                    
    blocker.after(0, 'block')                                                   
    nsproxy.shutdown(timeout=1.)                                                
                                                                                
    assert not addr.address.exists()
```

Reply:
Then there's something I don't understand about how the library works. I was thinking that if an agent was waiting to receive some message (therefore "blocked"), but before it came, a shutdown was issued, the agent would stop waiting ("unblock") and the execution would continue. Is there anything  wrong about my assumption?

Reply:
Well, as I said, if you are able to print the received message, then that means the reception was not interrupted by the shutdown. :wink:

In this case what is probably happening is that the shutdown has a timeout and, therefore, the name server might be waiting a couple of seconds for the agent to shutdown before actually interrupting it.

In the example I showed you (which is also a bit shorter), we are sure the agent will never unblock naturally (the block will last forever). So the only way to shut everything down is actually interrupting the agent.

Reply:
You are right, I tested my example with a longer waiting time and it worked (raised a `ConnectionClosedError` at the end, which was to be expected).

But then again,  your example works better. I'll add it and see if I need to change anything